pyproject_metadata/__init__.py

```diff
@@ -322,8 +369,21 @@ def write_to_rfc822(self, message: RFC822Message) -> None:  # noqa: C901
         message['Author-Email'] = self._email_list(self.authors)
         message['Maintainer'] = self._name_list(self.maintainers)
         message['Maintainer-Email'] = self._email_list(self.maintainers)
-        if self.license:
+
+        if isinstance(self.license, License):
             message['License'] = self.license.text
+            if self.metadata_version not in {'2.1', '2.2', '2.3'}:
+                warnings.warn(
+                    'Set license to an SPDX identifier for metadata >= 2.4',
```

Comment:
This should be pretty rare (the backend has to set metadata_version manually), and it could result in printing the warning twice, so I'd rather keep it here. The warning will happen a bit later in the process, but that's fine, warnings don't stop the compilation so being later might actually make it more visible. Errors stop the process, so that's different.